### PR TITLE
#164798263 Users can follow and unfollow each other

### PR DESCRIPTION
**Pull request by Genza999**
**What does this PR do?**
Creates functionality for users to follow or unfollow each other

**Description of tasks to be completed.**

- Creation of functionality of users to follow each other

- Creation of functionality of users to unfollow each other

- Creation of functionality to view a users followers

- Creation of functionality to view the users a user is following

- Ability to navigate to profiles of the followers and those being followed

**How can this be tested?**

- clone the repo

- Checkout ft-users-can-follow-eachother-164798263

- npm install && npm start

- log in with your account credentials

- Append an existing users username to the route to navigate to a users profile
   e.g. `https://localhost:8080/profile/{username}`

- Follow/Unfollow the user or click on the number of followers or following to view those users.

- Ability to follow/unfollow a user from their created articles

**Screenshots**

Screen Shot 2019-05-20 at 13 07 10

Screen Shot 2019-05-21 at 17 42 21

Screen Shot 2019-05-22 at 10 03 17

**Relevant Pivotal Tracker stories**

- [#164798263](https://www.pivotaltracker.com/story/show/164798263)

### CODE REVIEW

File: src/tests/actions/profileAction.test.js [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    ];
+
+    const editData = { profile: { full_name: "Ian Emma Ogwal" } };
+    return store.dispatch(editProfileAction(editData)).then(() => {
+      expect(store.getActions()).toEqual(expectedActions2);
+    });
+  });
+
+  // it("should get user following", () => {
+  //   const store = mockStore({});
```
Comment (patrickf949) [inline src/tests/actions/profileAction.test.js]: @Genza999 Great work sofar.
ensure to remove the commented out code

File: src/components/userProfile/EditProfile.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  handleNameChange(event) {
+    this.setState({ full_name: event.target.value });
+  }
+  handleBioChange(event) {
+    this.setState({ bio: event.target.value });
+  }
+  handleImageUpload(event) {
+    this.setState({ image: event.target.files });
+  }
+  handleSubmit(event) {
```
Comment (anyatibrian) [inline src/components/userProfile/EditProfile.js]: please @Genza999 thanks for your endurance, please remember to destructure your code

File: src/components/userProfile/EditProfile.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+      firebase
+        .storage()
+        .ref(`images/${files[0].name}`)
+        .getDownloadURL()
+        .then(url => {
+          this.setState({ uploadDone: true });
+          const image = {
+            image: url
+          };
+          this.props.editProfileAction(image);
```
Comment (anyatibrian) [inline src/components/userProfile/EditProfile.js]: @Genza999 Destructure

File: src/components/userProfile/EditProfile.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            <div className="row">
+              <div className="col-md-8">
+                <form role="form" method="post" onSubmit={this.handleSubmit}>
+                  <div className="form-group">
+                    <input
+                      id="full_name"
+                      type="text"
+                      placeholder="Enter your FullName"
+                      className="form-control"
+                      defaultValue={this.state.full_name}
```
Comment (anyatibrian) [inline src/components/userProfile/EditProfile.js]: @Genza999 please destructure

File: src/components/userProfile/EditProfile.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                    />
+                  </div>
+
+                  <div className="form-group">
+                    <input
+                      id="bio"
+                      type="text"
+                      placeholder="Enter a short Bio"
+                      className="form-control"
+                      defaultValue={this.state.bio}
```
Comment (anyatibrian) [inline src/components/userProfile/EditProfile.js]: @Genza999 Please Destructure

File: src/components/userProfile/EditProfile.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                    <img src={this.state.image} className="update-image" />
+                    <input
+                      type="file"
+                      onChange={event => this.uploadImage(event.target.files)}
+                      id="banyati"
+                      className="upload-button btn-outline-success"
+                    />
+                  </div>
+                </div>
+                <progress value={this.state.progress} max="100" />
```
Comment (anyatibrian) [inline src/components/userProfile/EditProfile.js]: @Genza999 please Destructure

File: src/components/userProfile/EditProfile.js (hunk 2 above)
Comment (Genza999) [reply src/components/userProfile/EditProfile.js]: Done!

File: src/components/userProfile/EditProfile.js (hunk 3 above)
Comment (Genza999) [reply src/components/userProfile/EditProfile.js]: Thanks. Done!

File: src/components/userProfile/EditProfile.js (hunk 4 above)
Comment (Genza999) [reply src/components/userProfile/EditProfile.js]: Done!

File: src/containers/SocialAuthContainer.js [hunk 7]
```diff
@@ -12,7 +12,8 @@ export class SocialAuthContainer extends Component {
   }
   componentWillReceiveProps(nextProps) {
     if (nextProps.socialAuthState.isAuthenticated) {
-      localStorage.setItem("token", nextProps.socialAuthState.payload);
+      sessionStorage.setItem("token", nextProps.socialAuthState.token);
+      sessionStorage.setItem("username", nextProps.socialAuthState.payload);
       this.props.history.push("/");
```
Comment (e-ian) [inline src/containers/SocialAuthContainer.js]: @Genza999 kindly destructure this.

File: src/components/userProfile/EditProfile.js (hunk 5 above)
Comment (Genza999) [reply src/components/userProfile/EditProfile.js]: Done!

File: src/tests/actions/profileAction.test.js (hunk 1 above)
Comment (Genza999) [reply src/tests/actions/profileAction.test.js]: Done. Thanks!

File: src/actions/profileActions.js [hunk 8]
```diff
@@ -1,30 +1,95 @@
...
+  FOLLOW_SUCCESS,
+  UNFOLLOW_FAILURE,
+  UNFOLLOW_SUCCESS,
+  FOLLOWERS_LIST,
+  FOLLOWING_LIST,
+  FOLLOWERS_LIST_FAILED,
+  FOLLOWING_LIST_FAILED
+} from "./types";
 
-export const getProfileAction = () => {
```
Comment (e-ian) [inline src/actions/profileActions.js:5]: @Genza999, why have you removed this. It is crushing the profile page for an authenticated user.

File: src/actions/profileActions.js [hunk 9]
```diff
@@ -1,30 +1,95 @@
...
+export const getUsersFollowing = username => {
   return async dispatch => {
     return axios
       .get(
-        `https://ah-backend-prime-staging.herokuapp.com/api/v1/profiles/${
-          sessionStorage.username
-        }`,
+        `https://ah-backend-prime-staging.herokuapp.com/api/v1/profiles/following/${username}`,
```
Comment (e-ian) [inline src/actions/profileActions.js:21]: @Genza999 why not separate concerns and have unfollow and follow separate. By editing this section of code you are breaking the individual user profile feature.

File: src/actions/profileActions.js [hunk 10]
```diff
@@ -1,30 +1,95 @@
...
             Authorization: "Bearer " + sessionStorage.token
           }
         }
       )
       .then(function(response) {
-        const data = response.data.profile;
+        const data = response.data.profiles;
         dispatch({
-          type: PROFILE_FETCHED,
+          type: FOLLOWERS_LIST,
```
Comment (e-ian) [inline src/actions/profileActions.js:58]: Same here, why did you change this action type.

File: src/actions/profileActions.js [hunk 11]
```diff
@@ -40,15 +107,125 @@ export const editProfileAction = data => {
...
+          payload: response.data.profile
+        });
+        toast.dismiss();
         toast.success("Profile updated successfully", {
           position: toast.POSITION.TOP_RIGHT,
           autoClose: 2000,
           hideProgressBar: false
         });
-        dispatch({
-          type: PROFILE_EDIT_SUCCESS,
```
Comment (e-ian) [inline src/actions/profileActions.js:49]: Where did this go?

File: src/components/navbar.js [hunk 12]
```diff
@@ -75,7 +79,7 @@ export class NavBarComponent extends React.Component {
               </li>
               <li className="nav-item">
                 <li className="nav-item">
-                  <Link className="nav-link" to="/profile">
```
Comment (e-ian) [inline src/components/navbar.js]: this link is supposed to route to the profiles page.

File: src/components/userProfile/EditProfile.js [hunk 13]
```diff
@@ -25,7 +25,7 @@ export class EditProfile extends Component {
   componentWillReceiveProps(nextProps) {
     if (nextProps.upDateSuccess === true) {
       const { history } = nextProps;
-      history.push("/profile");
+      history.push(`/profile/${sessionStorage.getItem("username")}`);
```
Comment (e-ian) [inline src/components/userProfile/EditProfile.js:28]: What is this?

**Review by e-ian — DISMISSED** [review]

File: src/components/userProfile/EditProfile.js (hunk 13 above)
Comment (Genza999) [reply src/components/userProfile/EditProfile.js:28]: Hey Emma, since you worked on getting a single users profile, getting other users profiles highly relied on your files as your task was oblivious of how to get other users profiles so I had to change it in a way to facilitate that. The session was included in the profiles routes in order to be able to navigate to other users profiles and not just your own.

**Review by anyatibrian — APPROVED** [review]
GREAT WORK @Genza999

**Review by e-ian — COMMENTED** [review]
image
@Genza999 in the above component the image of the profile has moved too close to the bio info. that's why when I edit my bio with more characters it overlaps the image as shown above.

**Comment by Genza999** [comment]
> image
> @Genza999 in the above component the image of the profile has moved too close to the bio info. that's why when I edit my bio with more characters it overlaps the image as shown above.

Thanks.Done!

**Review by e-ian — APPROVED** [review]
Great work @Genza999

**Comment by patrickf949** [comment]
Great Implementation @Genza999 
All the functionality works as expected
However, could you style this modal that appears when I click on followers on a user profile

If you give the image container a width this will look ok
Screenshot 2019-05-20 at 21 32 57

**Comment by Genza999** [comment]
> Great Implementation @Genza999
> All the functionality works as expected
> However, could you style this modal that appears when I click on followers on a user profile
> 
> If you give the image container a width this will look ok
> Screenshot 2019-05-20 at 21 32 57

Thanks alot! I have worked on it.

**Review by patrickf949 — APPROVED** [review]
Great work @Genza999  LGTM

File: src/tests/components/followUnfollowComponents.test.js [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+  });
+});
+
+describe("follower click", () => {
+  it("should click on a follower", () => {
+    const props = {
+      followers: ["david"]
+    };
+    const wrapper = mount(<FollowersList {...props} />);
+    wrapper.find("div#follower").simulate("click");
```
Comment (MuhanguziDavid) [inline src/tests/components/followUnfollowComponents.test.js:75]: you have to expect something in your test!!!

File: src/tests/components/followUnfollowComponents.test.js [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+  });
+});
+
+describe("following click", () => {
+  it("should click on a user  that is being followed", () => {
+    const props = {
+      following: ["david"]
+    };
+    const wrapper = mount(<FollowingList {...props} />);
+    wrapper.find("div#following").simulate("click");
```
Comment (MuhanguziDavid) [inline src/tests/components/followUnfollowComponents.test.js:86]: Expect something in all your tests

File: src/tests/components/followUnfollowComponents.test.js (hunk 14 above)
Comment (Genza999) [reply src/tests/components/followUnfollowComponents.test.js:75]: Thanks! I have added an expect for that test.

File: src/tests/components/followUnfollowComponents.test.js (hunk 15 above)
Comment (Genza999) [reply src/tests/components/followUnfollowComponents.test.js:86]: Thanks! I have also added an expect for this.

**Review by MuhanguziDavid — APPROVED** [review]
LGTM
